main.go

```diff
@@ -23,6 +26,42 @@ func main() {
 	mux.HandleFunc("/health", health.ServeHTTP)
 
 	loggedMux := loghandler.NewLogHandler(mux)
-	log.Printf("server is listening at %s", port)
-	log.Fatal(http.ListenAndServe(fmt.Sprintf(":%s", port), loggedMux))
+
+	log.Printf("server is listening at %s\n", port)
+
+	cfg.DebugContext = log.Printf
+
+	ctx := cfg.Context()
```

Comment:
i do not quite follow the need for custom context wrapper usage here. It adds extra complexity managing 'running' state, which i do not see the reason for.

Why we cannot just wait on channel? 

done := make(chan os.Signal, 1)
signal.Notify(done, os.Interrupt, syscall.SIGINT, syscall.SIGTERM)
<-done
call server.shutdown(here goes timeout context)

Reply:
Because proper interruptible context is not only used for awaiting for graceful shutdown. It is also used as a base context for every HTTP request (see http.Server struct). It is also used during request workflow to stop processing pipeline when context is interrupted/cancelled. It is also will be used for every Redis connection for the same purpose. It is also will be used with every Postgres connection.

In short, it is the only correct way to work with cancellation in Go.
 
The proposed implementation is not complex at all.

Reply:
In graceful shutdown server stops serving new requests, and completes all actively running requests. We should not cancel all outstanding requests. That's the thing - to wait until all of them served  - then die.

If instead we want to cancel all actively running requests we could use BaseContext with context.WithCancel. Standard contexts allow us to do that. 

@vlkolmakov I am not trying to

Reply:
@kepkap I don't understand what you're trying to say and how my approach contradicts the graceful shutdown workflow. Show me exact line number(s) where is my impl is incorrect.

contextWithCancel does not intercept OS signals.  Cancel func should be managed somewhere - this is what SignalContext is exactly doing.

And again, interruptible context should be an application level singleton, not only for HTTP server but for other users as well.

Reply:
ok cool, thanks for clarification.